tests/phpunit/TestFramework/Coverage/JUnit/JUnitTestCaseSorterTest.php

```diff
@@ -96,4 +103,133 @@ public function test_it_returns_unique_and_sorted_by_time_test_cases(): void
         $this->assertCount(3, $uniqueSortedFileNames);
         $this->assertSame('/path/to/test-file-3', $uniqueSortedFileNames[0]);
     }
+
+    public function test_it_has_correct_constants_for_bucket_sort(): void
+    {
+        $this->assertLessThan(
+            // Quicksort's average O(n log n)
+            JUnitTestCaseSorter::USE_BUCKET_SORT_AFTER * log(JUnitTestCaseSorter::USE_BUCKET_SORT_AFTER),
+            // Bucket Sort's average O(n + k)
+            JUnitTestCaseSorter::USE_BUCKET_SORT_AFTER + JUnitTestCaseSorter::BUCKETS_COUNT
+        );
+    }
+
+    /**
+     * @dataProvider locationsArrayProvider
+     */
+    public function test_it_sorts_correctly(array $uniqueTestLocations): void
+    {
+        $sortedTestLocations = iterator_to_array(JUnitTestCaseSorter::bucketSort($uniqueTestLocations));
+        $this->assertTrue(self::isOrderConstraintsValid($sortedTestLocations), 'Bucket sort failed order check');
+    }
+
+    /**
+     * Sanity check
+     *
+     * @dataProvider locationsArrayProvider
+     */
+    public function test_quicksort_sorts_correctly(array $uniqueTestLocations): void
+    {
+        self::quicksort($uniqueTestLocations);
+
+        $this->assertTrue(self::isOrderConstraintsValid($uniqueTestLocations), 'Quicksort failed order check');
+    }
+
+    /**
+     * @dataProvider locationsArrayProvider
+     */
+    public function test_it_sorts_faster_than_quicksort(array $uniqueTestLocations): void
+    {
+        if (extension_loaded('xdebug') || PHP_SAPI === 'phpdbg') {
+            $this->markTestSkipped('Benchmarks under xdebug or phpdbg are brittle');
+        }
+
+        if (self::isOrderConstraintsValid($uniqueTestLocations) === true) {
+            // Ignore silently as to not pollute to the log.
+            $this->addToAssertionCount(1);
+
+            return;
+        }
+
+        $tries = 100;
+
+        // Benchmark bucket sort
+        $totalBucketSort = 0;
+
+        for ($i = 0; $i < $tries; ++$i) {
+            $start = microtime(true);
+            iterator_to_array(JUnitTestCaseSorter::bucketSort($uniqueTestLocations));
+            $totalBucketSort += microtime(true) - $start;
+        }
+
+        // Benchmark quicksort
+        $totalQuickSort = 0;
+
+        for ($i = 0; $i < $tries; ++$i) {
+            $start = microtime(true);
+            // Updates by reference
+            $locationsCopy = $uniqueTestLocations;
+            self::quicksort($locationsCopy);
+            $totalQuickSort += microtime(true) - $start;
+        }
+
+        $this->assertGreaterThanOrEqual(0.01, abs($totalQuickSort - $totalBucketSort));
+    }
+
+    public static function locationsArrayProvider(): iterable
```

Comment:
```suggestion
    /**
     * @return iterable<TestLocation>
     */
    public static function locationsArrayProvider(): iterable
```

Reply:
That's iterable of array of array of locations. Not sure if it makes sense to document this since we're not using this method directly.

Reply:
no you're right, we don't for providers...